File: Difficulty/Hard/[164] Maximum Gap/solution.cpp
/*
 * @lc app=leetcode id=164 lang=cpp
 *
 * [164] Maximum Gap
 */

// @lc code=start
class Solution {
public:
    int maximumGap(vector<int>& nums) {
        if (nums.size() < 2) return 0;
        auto lu = minmax_element(nums.begin(), nums.end());
        int l = *lu.first, u = *lu.second;
        int gap = ceil(double(u - l) / (nums.size() - 1));
        vector<int> bucketsMIN(nums.size() - 1, INT_MAX);
        vector<int> bucketsMAX(nums.size() - 1, INT_MIN);
        for (auto n : nums) {
            if (n == l || n == u)
                continue;
            int idx = (n - l) / gap;
            bucketsMIN[idx] = min(n, bucketsMIN[idx]);
            bucketsMAX[idx] = max(n, bucketsMAX[idx]);
        }
        int maxGap = INT_MIN;
        int pre = l;
        for (int i = 0; i < nums.size() - 1; i++) {
            if (bucketsMIN[i] == INT_MAX && bucketsMAX[i] == INT_MIN)
                continue;
            maxGap = max(maxGap, bucketsMIN[i] - pre);
            pre = bucketsMAX[i];
        }
        return max(maxGap, u - pre);
    }
};
// @lc code=end

File: Difficulty/Hard/[164] Maximum Gap/solution.py
class Solution:
    def maximumGap(self, nums: List[int]) -> int:
        if len(nums) < 2: return 0
        l, u = 2 ** 31 - 1, -2 ** 31
        for num in nums:
            l = min(l, num)
            u = max(u, num)
        gap = math.ceil((u - l) / (len(nums) - 1))
        bucketsMIN, bucketsMAX = [2 ** 31 - 1] * (len(nums) - 1), [-2 ** 31] * (len(nums) - 1)
        for num in nums:
            if num == l or num == u:
                continue
            idx = int((num - l) / gap)
            bucketsMIN[idx] = min(num, bucketsMIN[idx])
            bucketsMAX[idx] = max(num, bucketsMAX[idx])
        maxGap, pre = -2 ** 31, l
        for i in range(len(nums) - 1):
            if bucketsMIN[i] == 2 ** 31 - 1 and bucketsMAX[i] == -2 ** 31:
                continue
            maxGap = max(maxGap, bucketsMIN[i] - pre)
            pre = bucketsMAX[i]
        return max(maxGap, u - pre)